File: extension/src/shared/classifier.ts
export interface ClassificationResponse {
  label: string;
  score: number;
  modelVersion: string;
  source: "onnx" | "keyword";
}

type Signal = {
  label: string;
  pattern: RegExp;
  weight: number;
};

const SIGNALS: Signal[] = [
  { label: "urgent", pattern: /\b(urgent|asap|immediately|eod|blocked|critical|important)\b/i, weight: 0.36 },
  { label: "follow-up", pattern: /\b(following up|follow up|checking in|reminder|gentle reminder|waiting)\b/i, weight: 0.3 },
  { label: "action-required", pattern: /\b(please|can you|could you|need you to|submit|confirm|send|review)\b/i, weight: 0.28 },
  { label: "opportunity", pattern: /\b(internship|placement|tnp|training and placement|career opportunity|hiring|apply|campus drive)\b/i, weight: 0.42 },
  { label: "work", pattern: /\b(project|meeting|deadline|review|proposal|invoice|client|report)\b/i, weight: 0.26 },
  { label: "finance", pattern: /\b(invoice|payment|receipt|billing|refund|paid|amount due)\b/i, weight: 0.25 },
  { label: "personal", pattern: /\b(family|dinner|birthday|weekend|trip|holiday)\b/i, weight: 0.22 },
  { label: "spam", pattern: /\b(winner|free money|claim now|limited offer|act now|prize)\b/i, weight: 0.34 },
  { label: "newsletter", pattern: /\b(newsletter|unsubscribe|weekly update|digest|read more)\b/i, weight: 0.24 }
];

const BASE_MODEL_VERSION = "keyword-classifier-phase-2";
const ONNX_MODEL_PATH = "models/classifier/model.onnx";

let onnxAvailability: Promise<boolean> | undefined;

export async function classifyEmailClient(emailText: string): Promise<ClassificationResponse> {
  if (await hasBundledOnnxModel()) {
    // The ONNX model slot is intentionally detected before fallback so a trained
    // classifier can be wired in without changing the popup/background flow.
    return classifyWithKeywordFallback(emailText, "onnx-ready-keyword-fallback");
  }

  return classifyWithKeywordFallback(emailText, BASE_MODEL_VERSION);
}

async function hasBundledOnnxModel(): Promise<boolean> {
  onnxAvailability ??= checkOnnxModel();
  return onnxAvailability;
}

async function checkOnnxModel(): Promise<boolean> {
  if (!globalThis.chrome?.runtime?.getURL) return false;

  try {
    const url = chrome.runtime.getURL(ONNX_MODEL_PATH);
    const response = await fetch(url, { method: "HEAD" });
    return response.ok;
  } catch {
    return false;
  }
}

function classifyWithKeywordFallback(emailText: string, modelVersion: string): ClassificationResponse {
  const normalized = emailText.replace(/\s+/g, " ").trim();
  const scores = new Map<string, number>();

  for (const signal of SIGNALS) {
    if (signal.pattern.test(normalized)) {
      scores.set(signal.label, (scores.get(signal.label) ?? 0.18) + signal.weight);
    }
  }

  if (hasInstitutionalSignal(normalized)) {
    scores.set("opportunity", (scores.get("opportunity") ?? 0.18) + 0.42);
    scores.set("work", (scores.get("work") ?? 0.18) + 0.18);
    if (scores.has("newsletter")) {
      scores.set("newsletter", (scores.get("newsletter") ?? 0) * 0.35);
    }
  }

  if (scores.size === 0) {
    return {
      label: normalized.length > 900 ? "newsletter" : "low-priority",
      score: normalized.length > 900 ? 0.68 : 0.62,
      modelVersion,
      source: "keyword"
    };
  }

  const [label, score] = [...scores.entries()].sort((a, b) => b[1] - a[1])[0];
  return {
    label,
    score: Math.min(0.95, Math.max(0.55, score)),
    modelVersion,
    source: "keyword"
  };
}

function hasInstitutionalSignal(text: string): boolean {
  return /\b(tnp|training and placement|placement cell|career development|iiit|iit|nit|university|college|\.edu\b|\.ac\.in\b|@[^@\s]+\.ac\.in\b)\b/i.test(text);
}

File: backend/app/services/classifier.py
import re
from dataclasses import dataclass


@dataclass(frozen=True)
class Classification:
    label: str
    score: float
    model_version: str


@dataclass(frozen=True)
class Signal:
    label: str
    pattern: re.Pattern[str]
    weight: float


class HeuristicEmailClassifier:
    model_version = "heuristic-classifier-v3"

    def __init__(self) -> None:
        self.signals = [
            Signal("urgent", re.compile(r"\b(urgent|asap|immediately|eod|blocked|critical|important)\b", re.I), 0.36),
            Signal("follow-up", re.compile(r"\b(following up|follow up|checking in|reminder|gentle reminder|waiting)\b", re.I), 0.3),
            Signal("action-required", re.compile(r"\b(please|can you|could you|need you to|submit|confirm|send|review)\b", re.I), 0.28),
            Signal("opportunity", re.compile(r"\b(internship|placement|tnp|training and placement|career opportunity|hiring|apply|campus drive)\b", re.I), 0.42),
            Signal("work", re.compile(r"\b(project|meeting|deadline|review|proposal|invoice|client|report)\b", re.I), 0.26),
            Signal("finance", re.compile(r"\b(invoice|payment|receipt|billing|refund|paid|amount due)\b", re.I), 0.25),
            Signal("personal", re.compile(r"\b(family|dinner|birthday|weekend|trip|holiday)\b", re.I), 0.22),
            Signal("spam", re.compile(r"\b(winner|free money|claim now|limited offer|act now|prize)\b", re.I), 0.34),
            Signal("newsletter", re.compile(r"\b(newsletter|unsubscribe|weekly update|digest|read more)\b", re.I), 0.24),
        ]

    def classify(self, email_text: str) -> Classification:
        normalized = " ".join(email_text.split())
        scores: dict[str, float] = {}

        for signal in self.signals:
            if signal.pattern.search(normalized):
                scores[signal.label] = scores.get(signal.label, 0.18) + signal.weight

        if self._has_institutional_signal(normalized):
            scores["opportunity"] = scores.get("opportunity", 0.18) + 0.42
            scores["work"] = scores.get("work", 0.18) + 0.18
            if "newsletter" in scores:
                scores["newsletter"] *= 0.35

        if not scores:
            if len(normalized) > 900:
                return Classification("newsletter", 0.68, self.model_version)
            return Classification("low-priority", 0.62, self.model_version)

        label, score = sorted(scores.items(), key=lambda item: item[1], reverse=True)[0]
        return Classification(label, min(0.95, max(0.55, score)), self.model_version)

    def _has_institutional_signal(self, text: str) -> bool:
        return bool(
            re.search(
                r"\b(tnp|training and placement|placement cell|career development|iiit|iit|nit|university|college|"
                r"\.edu\b|\.ac\.in\b|@[^@\s]+\.ac\.in\b)\b",
                text,
                re.I,
            )
        )
